torch/csrc/serialization.cpp

```diff
@@ -233,27 +234,22 @@ void THPStorage_writeFileRaw(
   int64_t numel = size_bytes / element_size;
   if (self->device_type() == at::kCPU) {
     data = self->data<uint8_t>();
-#if defined(USE_CUDA) && defined(TORCH_HIP_VERSION) && \
-    (TORCH_HIP_VERSION >= 301)
-  } else if (self->device_type() == at::kCUDA) {
-    cpu_data = std::unique_ptr<char[]>(new char[size_bytes]);
-    data = (uint8_t*)cpu_data.get();
-    C10_CUDA_CHECK(hipMemcpyWithStream(
-        data,
-        self->data<uint8_t>(),
-        size_bytes,
-        cudaMemcpyDeviceToHost,
-        c10::hip::getCurrentHIPStreamMasqueradingAsCUDA()));
-#elif defined(USE_CUDA)
-  } else if (self->device_type() == at::kCUDA) {
+  } else {
+    // Here we use a dispatched copy for all non-CPU device.
+    auto device_tensor = at::from_blob(
+        self->data<void>(),
+        {size_bytes},
+        {1},
+        NULL,
+        at::device(self->device()).dtype(c10::kByte),
+        {self->device()});
     cpu_data = std::unique_ptr<char[]>(new char[size_bytes]);
+    auto cpu_tensor = at::from_blob(
+        (void*)cpu_data.get(),
+        {size_bytes},
+        at::device(at::kCPU).dtype(c10::kByte));
+    cpu_tensor.copy_(device_tensor);
```

Comment:
This seems a bit more circuitous than is necessary. Since you're doing a dispatched copy, the operator can take care of allocating a CPU tensor for you. So you can just directly convert `device_tensor` to CPU, and then pull out the data pointer directly, no need to hand allocate `cpu_data` buffer anymore.

Reply:
Done. It looks more clear.
@ezyang Thanks, any more comments?